2단계 - 지하철 노선 기능

chj1768 opened the pull request:
안녕하세요 리뷰어님. 🙇 
2단계 PR 요청 드립니다. 
잘 부탁 드립니다.

chj1768 committed:
docs: 2단계 문서 작성

chj1768 committed:
feat: 실패하는 인수테스트 작성 - 지하철노선 생성

chj1768 committed:
feat: 노선 생성, 조회 로직 작성

chj1768 committed:
feat: 테스트 수정, LineRepositoryTest 추가

chj1768 committed:
feat: Acceptance Test 용 빈 추가, 프로파일 설정

chj1768 committed:
feat: 노선 AcceptanceTest 내 DatabaseCleaner 적용, 단일 조회, 목록 테스트 추가

chj1768 committed:
feat: 실패하는 생성, 수정, 단일 조회 테스트 추가

chj1768 committed:
feat: 실패하는 생성, 수정, 단일 조회 구현 로직 추가

chj1768 committed:
feat: 노선 인수테스트 수정

chj1768 committed:
refactor: StationAcceptanceTest 수정 - DatabaseCleaner 적용(테스트 격리), 검증시 하드코딩된 값 사용하도록 변경

chj1768 committed:
docs: 요구사항 완료 마킹

chj1768 committed:
refactor: 불필요 에러핸들러 제거

chae-yh commented on README.md:
```diff
@@ -55,20 +55,83 @@ This project is [MIT](https://github.com/next-step/atdd-subway-admin/blob/master
+- [X] 인수 조건을 검증하는 인수 테스트를 작성하세요
 
-* 인수 테스트의 재사용성과 가독성, 그리고 빠른 테스트 의도 파악을 위해 인수 테스트를 리팩터링 하세요.
+### 프로그래밍 요구사항
```
README 파일을 자세히 작성해주시는게 인상깊네요 👍 
기능 구현 목록을 구체적으로 작성해보는 것은 어떨까요...?
ex) 
- 노선 생성 api 를 작성한다
   - 지하철 노선은 여러개의 지하철역을 가질 수 있다
   - 지하철역은 section 이라는 객체를 통해 소유한다

chj1768 replied:
작성해주신 예시 처럼 
제가 작업했던 스텝을 그룹화해서 명시해보았습니다. 
[06d7c56](https://github.com/next-step/atdd-subway-admin/pull/696/commits/06d7c56fd72d6ff8d81434afdf4b5b1752c2bbc5)

chae-yh commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.stream.Collectors;
+
+@Service
+@Transactional(readOnly = true)
```
여기에 Transactional을 걸고 readOnly는 해당 메소드에만 붙이는 건 어떨까요...?
Transactional이 더 많기도 하고 readOnly가 특수한 경우라서 해당 하는 메소드에만 작성하는 것은 어떨까요...?

chj1768 replied:
[8fc51a5](https://github.com/next-step/atdd-subway-admin/pull/696/commits/8fc51a56b13f7553ea156cc1ce5beb261843d65f) 에서 변경해주었습니다. :-)

chae-yh commented on src/main/java/nextstep/subway/NotFoundException.java:
```diff
@@ -0,0 +1,4 @@
+package nextstep.subway;
+
+public class NotFoundException extends RuntimeException {
```
해당 형태면 혹시 NoSuchElementException를 써도 되지 않을까요...?

chj1768 replied:
네. [633322c](https://github.com/next-step/atdd-subway-admin/pull/696/commits/633322c65838cbfd66ce22553d205387d54c02c1) 에서 변경해보았습니다. 
(NosuchElementException 은 리스트 같은 순회 가능한 형식 안에서 사용하는 예외로 생각했었는데, 꼭 그렇지는 않군요!!)

chae-yh commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,60 @@
+
+    @Transactional
+    public LineResponse saveLine(LineRequest lineRequest) {
+        Station upStation = stationRepository.findById(lineRequest.getUpStationId()).orElseThrow(NotFoundException::new);
```
여기도 method extract 해줄 수 있을 것 같습니다!

chj1768 replied:
[b8f4988](https://github.com/next-step/atdd-subway-admin/pull/696/commits/b8f498898ecc89e10890f7656a5544cefb9f6ca3) 에서 추출해주었습니다.

chae-yh commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.List;
+import java.util.Objects;
+
+@DynamicUpdate
```
혹시 해당 annotation을 사용하는 이유가 있을까요...? 맞다 / 틀리다의 부분이 아니라 궁금하여 문의 드립니다!

chj1768 replied:
@chae-yh 
지하철 호선 엔티티 변경 시 쿼리에서 **수정 대상 필드만 업데이트 시키게끔 동작하려고** 추가하게 되었습니다. 
DynamicUpdate 추가 전 
![image](https://user-images.githubusercontent.com/15382538/171417183-4595f900-0dff-419d-9cbf-0133dca78617.png)
DynamicUpdate 추가 후
![image](https://user-images.githubusercontent.com/15382538/171417230-2420c17a-b07e-4355-b258-b183afa588e7.png)

chae-yh commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -0,0 +1,79 @@
+        return color;
+    }
+
+    public List<Station> getStations() {
```
해당 형태면 예를 들어, 2호선에는 상행 종점 / 하행 종점만 있을 것 같습니다. 2호선에는 더 많은 역이 들어갈 수 있는 구조여야 하지 않을까요...?

chj1768 replied:
@chae-yh 
<img width="937" alt="image" src="https://user-images.githubusercontent.com/15382538/171417863-902b1707-9111-4608-8cd2-bcee1f2411eb.png">
요 명세를 보고,
현 단계에서는 노선 생성시 상행종점역과 하행종점역만을  포함하여 등록한다.
라고 생각했습니다. 
따라서 두가지 역만 노선에 포함 되고, 조회시에는 두 역을 리스트로 뷰잉하는 구조로 만들게 되었습니다. (리스트 부분은 API 응답 명세를 참고하여 만들게 되었습니다.)
혹시 놓친 부분이 있을까요? 🙏

chae-yh replied:
넵 말씀하신대로 line을 만들 때는 상행 종점과 하행 종점을 등록해야합니다. 추가로, 현재 station을 추가로 매핑하는 구조는 없습니다만, 현재 만들어주신 구조라면 line에 station이 2개 이상 추가할 수 없는 구조가 아닌가하여 말씀드렸습니다!
다음 미션이 station을 line에 추가로 매핑해주는 작업이 될 텐데 여기서 line 내에 station을 list로 가지는 부분의 작업이 진행될 것으로 기대합니다!

chae-yh commented on src/main/java/nextstep/subway/domain/LineRepository.java:
```diff
@@ -0,0 +1,20 @@
+
+public interface LineRepository extends JpaRepository<Line, Long> {
+    @Override
+    @EntityGraph(attributePaths = { "upStation", "downStation" })
```
@EntityGraph 를 사용하는 경우는 미션 중에는 처음보네요 👍

chj1768 replied:
역 정보까지 같이 들고올 수 있도록 페치 조인 방식 대신으로 한번 사용해보았습니다. (_ _)

chae-yh commented on src/main/java/nextstep/subway/domain/LineRepository.java:
```diff
@@ -0,0 +1,20 @@
+    @Override
+    @Modifying
+    @Query("delete from Line a where a.id = :id")
+    void deleteById(@Param("id") Long id);
```
JPA에서 제공해주는 delete()를 사용하지 않은 이유를 여쭤봐도 될까요..? 이부분도 맞고 틀림의 문제는 아니고 궁금하여 여쭤봅니다!

chj1768 replied:
@chae-yh 
디폴트 구현에서는
영속성 컨텍스트에서 제거하기 전에 해당 엔티티가 실제로 존재하는지 확인하는 작업이 포함되어 있는 걸로 알고 있었습니다.
select + delete 로 쿼리가 2번 수행되는걸 확인해서 
delete 쿼리 1번만 수행하도록 기존 구현 오버라이딩에 JPQL 을 적용했습니다.

chae-yh commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -0,0 +1,55 @@
+
+    @DeleteMapping("/{id}")
+    public void deleteLine(@PathVariable Long id) {
+        lineService.deleteLine(id);
```
void로 아무런 response도 안주기보다는, 삭제 되었음을 나타내는 response를 주는 것은 어떨까요...?

chj1768 replied:
[fd66dd9](https://github.com/next-step/atdd-subway-admin/pull/696/commits/fd66dd9709e9469abd8344372d377492ee391e68) 네, 여기서 역 삭제 응답과 동일한 204 응답으로 표현하도록 변경해봤습니다.

chae-yh commented on src/test/java/nextstep/subway/domain/LineRepositoryTest.java:
```diff
@@ -0,0 +1,44 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DataJpaTest
+class LineRepositoryTest {
```
슬라이스 테스트 좋습니다 👍

chae-yh commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,194 @@
+
+    @AfterEach
+    public void tearDown() {
+        databaseCleaner.clean();
```
테스트 격리 좋습니다 👍

chae-yh commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    private ExtractableResponse<Response> getOne(Long id) {
+        return RestAssured.given().log().all().when().get("/lines/{id}", id).then().log().all().extract();
```
이거는 개행을 해주셔도 좋을 것 같습니다!
```
return RestAssured.given().log().all()
    .when().get("/lines/{id}", id)
    .then().log().all()
    .extract();
```

chj1768 replied:
[b180810](https://github.com/next-step/atdd-subway-admin/pull/696/commits/b18081042dcd85d69ea87f88317ba85a62a1a626) 
네, 요 커밋에서 given/when/then 단위로 개행처리 했습니다. (가독성 향상에 아주 좋네요.)

chae-yh commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,194 @@
+        assertThat(actual.jsonPath().getString("name")).isEqualTo("신분당선");
+    }
+
+    private ExtractableResponse<Response> getOne(Long id) {
```
RestAssured 메소드 분리 잘해주셨습니다!

chae-yh commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,194 @@
+
+    private ExtractableResponse<Response> createLine(
+            String name, String color, long upStationId, long downStationId, long distance) {
+        Map<String, String> params = new HashMap<>();
```
Request를 써볼 수도 있을 것 같습니다!

chj1768 replied:
[4228c39](https://github.com/next-step/atdd-subway-admin/pull/696/commits/4228c398433a8bdbc04a32bba7f211e1f0b2bffa) 네. 여기에 포함되어 개선하였습니다. 
맵 대신 서비스에서 사용하는 DTO 타입을 사용하여 테스트에서 생성요청 하도록 변경했습니다.

chae-yh commented on src/test/java/nextstep/subway/station/StationAcceptanceTest.java:
```diff
@@ -23,13 +26,21 @@ public class StationAcceptanceTest {
+
     @BeforeEach
     public void setUp() {
         if (RestAssured.port == RestAssured.UNDEFINED_PORT) {
```
이 부분은 LineAcceptanceTest에도 나오는 등 반복적으로 나오는데 공통으로 추출해낼 수 있지 않을까요...?

chj1768 replied:
[4228c39](https://github.com/next-step/atdd-subway-admin/pull/696/commits/4228c398433a8bdbc04a32bba7f211e1f0b2bffa) 에서 공통으로 사용하게 되는 인수 테스트용 클래스 요소들을 분리해보았습니다.

chae-yh commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -0,0 +1,79 @@
+    protected Line() {
+    }
+
+    public Line(String name, String color, Station upStation, Station downStation, Long distance) {
```
distance 등의 validation이 추가될 수 있지 않을까요...?

chj1768 replied:
네, 추가될 수 있다고 생각합니다. LineRequest, LineUpdateRequest 와 같은 DTO 에서 validation 을 수행하도록 하려고 합니다. (각각의 값에 대한 특정한 validation 조건이 없는것 같아 현재는 별도의 로직은 없습니다. )

chae-yh replied:
👍 
validation관련해서는, 예를들어 distance에 음수의 값이 들어와도 해당 Line이 만들어져야 할까요...?

chj1768 replied:
@chae-yh 

> distance에 음수의 값이 들어와도 해당 Line이 만들어져야 할까요...?

앗 넵넵 다음 PR 드릴때 validation 부분 체크해보겠습니다. 🙇

chae-yh commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -0,0 +1,79 @@
+        return Arrays.asList(this.upStation, this.downStation);
+    }
+
+    public void update(LineUpdateRequest request) {
```
도메인이 Request를 바로 알기보다는 값만 받아오면 어떨까요...?
```suggestion
    public void update(String name, String color) {
```
Request 같은 dto의 변경이 도메인의 변경에 영향을 주는 구조는 피하는 경향이 있는 것 같습니다!

chj1768 replied:
네. [927c2de](https://github.com/next-step/atdd-subway-admin/pull/696/commits/927c2ded78ac9bb925a8280d80da09bb4116078c) 에서 값만 받아오도록 변경해보았습니다.

chae-yh commented on src/test/java/nextstep/subway/line/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,194 @@
+    void getAll() {
+        // given
+        ExtractableResponse<Response> response1 =
+                createLine("신분당선", "bg-red-600", upStation.getId(), downStation.getId(), 10);
```
신분당선은 반복적으로 나오는데, BeforeEach로 미리 만들어 두고 사용할 수 있지 않을까요...?

chj1768 replied:
[4228c39](https://github.com/next-step/atdd-subway-admin/pull/696/commits/4228c398433a8bdbc04a32bba7f211e1f0b2bffa) 에서 포함하여 개선하였습니다. 
해당 request 를 멤버필드로  선언하고 필요시 생성요청하여 사용할 수 있도록 했어요.

chae-yh reviewed (changes requested):
현지님 안녕하세요! 2번째 단계도 코드 잘 작성해주셨네요 👍 
다만 고민해보시면 좋을만한 내용이 있어 코멘트 드렸으니 확인 부탁드립니다

감사합니다 : )

chj1768 committed:
docs: 요구사항 상세 그룹화하여 명시

chj1768 committed:
refactor: 읽기 전용 트랜젝션 어노테이션을 메서드에서 별도 명시하도록 순위변경

chj1768 committed:
refactor: NosuchException 으로 NotFoundException 교체

chj1768 committed:
refactor: station findById method extract

chj1768 committed:
refactor: void 타입 응답 명시적으로 ResponseEntity 리턴하도록 변경

chj1768 committed:
refactor: RestAssured 사용시 given/when/then 단위 개행 적용

chj1768 committed:
refactor: Request 같은 dto의 변경이 도메인의 변경에 영향을 주는 구조를 값만 받아오도록 변경

chj1768 committed:
refactor: 인수 테스트 클래스 개선
- 인수 테스트 공통 역할 클래스로 추출하여 각 인수테스트에서 확장하여 사용하도록 분리
- LineAcceptanceTest 내 라인 생성 request 공통 객체 필드로 선언하여 추출
- 라인 생성 요청시 request body 를 map 대신 서비스에서 사용하는 LineRequest DTO 사용하여 생성하도록 테스트 변경

chj1768 committed:
refactor: 지하철 호선 삭제 시 204 주도록 수정

chae-yh reviewed (approved):
현지님 안녕하세요! 피드백 잘 반영해주셨습니다 👍 
고민해 주셨으면 하는 부분이 있어 소소하게 코멘트 남겨드렸지만 이정도면 미션 목표는 완수되었다고 생각되기에 merge 하겠습니다!

다음 미션 진행 부탁드립니다!
